src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs

```diff
@@ -3327,4 +3344,404 @@ internal void SetDerivedNames(string protocol, string serverName)
+
+        /// <summary>
+        /// Provides the pipe name in case of Named Pipes
+        /// </summary>
+        internal string PipeName { get; private set; }
+
+        /// <summary>
+        /// Provides the HostName to connect to in case of Named pipes Data Source
+        /// </summary>
+        internal string PipeHostName { get; private set; }
+
+        private string _workingDataSource;
+        private string _dataSourceAfterTrimmingProtocol;
+
+        internal bool IsBadDataSource { get; private set; } = false;
+
+        internal bool IsSsrpRequired { get; private set; } = false;
+
+        private DataSource(string dataSource)
+        {
+            // Remove all whitespaces from the datasource and all operations will happen on lower case.
+            _workingDataSource = dataSource.Trim().ToLowerInvariant();
+
+            int firstIndexOfColon = _workingDataSource.IndexOf(SemiColon);
+
+            PopulateProtocol();
+
+            _dataSourceAfterTrimmingProtocol = (firstIndexOfColon > -1) && _connectionProtocol != Protocol.None
+                ? _workingDataSource.Substring(firstIndexOfColon + 1).Trim() : _workingDataSource;
+
+            Debug.Assert(_dataSourceAfterTrimmingProtocol.Contains(Slash), "pipe paths only allow back slashes");
+        }
+
+        private void PopulateProtocol()
+        {
+            string[] splitByColon = _workingDataSource.Split(SemiColon);
+
+            if (splitByColon.Length <= 1)
+            {
+                _connectionProtocol = Protocol.None;
+            }
+            else
+            {
+                // We trim before switching because " tcp : server , 1433 " is a valid data source
+                switch (splitByColon[0].Trim())
+                {
+                    case TdsEnums.TCP:
+                        _connectionProtocol = Protocol.TCP;
+                        break;
+                    case TdsEnums.NP:
+                        _connectionProtocol = Protocol.NP;
+                        break;
+                    case TdsEnums.ADMIN:
+                        _connectionProtocol = Protocol.Admin;
+                        break;
+                    default:
+                        // None of the supported protocols were found. This may be a IPv6 address
+                        _connectionProtocol = Protocol.None;
+                        break;
+                }
+            }
+        }
+
+        // LocalDbInstance name always starts with (localdb)
+        // possible scenarios:
+        // (localdb)\<instance name>
+        // or (localdb)\. which goes to default localdb
+        // or (localdb)\.\<sharedInstance name>
+        internal static string GetLocalDBInstance(string dataSource, out bool error)
+        {
+            string instanceName = null;
+
+            // ReadOnlySpan is not supported in netstandard 2.0, but installing System.Memory solves the issue
+            ReadOnlySpan<char> input = dataSource.AsSpan().TrimStart();
```

Comment:
Thanks for using spans, almost always better for perf 😁

Reply:
This is c/p of a class from netcore path. This needs some refactoring and unification as it's useful for both necore and netfx.